### DISCUSSION TITLE
filenameEncoding

### DISCUSSION THREAD

**Discussion by nenge123** [gildas-lormeau/zip.js#351] [General]
can you see auto decode name

zip.js line 3068
const commentEncoding = getOptionValue$1(zipReader, options, "commentEncoding");
```js
const arrayToString = (t,u)=>{
    let m='utf8',w='cp437',k=(u||w).toLowerCase(),text =new TextDecoder(m).decode(t);
    return t.byteLength!=new TextEncoder().encode(text).byteLength?(k==w?decodeCP437(t):new TextDecoder(k).decode(t)):text;
};
const [filename, comment] = await Promise.all([
    arrayToString(fileEntry.rawFilename,filenameEncoding),
    decodeText(fileEntry.rawComment, fileEntry.commentUTF8 ? CHARSET_UTF8 : commentEncoding || CHARSET_CP437)
]);			
```

**Comment by gildas-lormeau** [comment]
It looks like the filename is encoded in UTF-8 but the bit 11 of the "general purpose bit flag" is not set to 1 (see 4.4.4 here: https://pkware.cachefly.net/webdocs/APPNOTE/APPNOTE-6.3.9.TXT). That's why `filenameUTF8` is set to `false`. From my point of view, the zip file you're testing does not follow the specification. It should also be reproducible on any OS.

**Reply by nenge123** [reply]
here is error
you set default filenameEncoding is 'GBK'
----


![image](https://user-images.githubusercontent.com/16117315/184460037-08864afd-2951-4524-9d4d-7166035880c4.jpeg)

**Reply by nenge123** [reply]
When I set "filenameEncoding", please consider using the length to determine whether it is utf8? Decode?


let raw = entry.rawFilename

raw.byteLength != new TextEncode().encode(new TextDecoder('gbk').decode(raw)).byteLength

**Comment by nenge123** [comment]


**Reply by gildas-lormeau** [reply]
I'm sorry but I don't understand what you wrote.

**Comment by nenge123** [comment]
please see my top post for filename decode!thanks

**Comment by nenge123** [comment]
and then plase delete filenameEncoding default value "gbk"

**Reply by gildas-lormeau** [reply]
I'm sorry but I still don't understand the problem. Moreover, the code you are referring to in the first post does not exist, see below. 

https://github.com/gildas-lormeau/zip.js/blob/238f901a6a0081f166d66515658a7f76546958ba/dist/zip.js#L3068

https://github.com/gildas-lormeau/zip.js/blob/238f901a6a0081f166d66515658a7f76546958ba/dist/zip.js#L3091-L3094

Maybe the best is that you express the problem *clearly* in your native language and I will do the translation on my side.

**Comment by gildas-lormeau** [comment]
Check the encoding of your `zip.js` file, it must be served in UTF-8. see https://github.com/gildas-lormeau/zip.js/issues/352

**Comment by nenge123** [comment]
- 在你的zip.js,line:2969行设置默认编码GBK
这应该不太好吧？

- 假设非UTF8字符，它解码后重新转回UTF8编码，那么长度肯定是不一致的。


- 浏览器解压大于32MB，非常慢。100MB差不多几分钟！

- 另外每次解压都会创建多个内容一样的BLOB:http:// (zip.js line:117)连接。为何不复用连接，如果是worker也可以暂时不关闭，直到使用zipReader.close()

**Comment by gildas-lormeau** [comment]
The original issue in this discussion is fixed in the version v2.6.13 I just published. 

Please create a new discussion for each different type of issue/suggestion. 

It shouldn't take minutes to decompress 100 MB. On my machine, it takes usually less than 10 seconds.

I don't understand the optimization related to the Blob URI used by web workers you suggest. The line 117 is executed one time only.

**Comment by nenge123** [comment]
文件智能编码问题 :fileEncoding

实际文件:cp437 [148]
utf8 [192,148]
gbk [63]实际上不存在

let dd = (e,s)=>new TextDecode(e).decode(new Uint8Array(s));
let ee = e=>new TextEncoder().encode(e);

ee(dd("utf8",[148])); 基础编码，长度3
ee(dd("gbk",[148])); 自定义编码，长度3
ee(cp437[148]);默认编码，长度2


实际文件:gbk [239,210] 我
utf8 230,136,145

ee(dd("utf8",[239，210]));基础编码 長度6
ee(dd("gbk",[239，210]));我的编码，長度3
ee(cp437[239]+cp437[210]);默认编码，長度6
这里utf8与cp437长度一样，但是解码字符不一致！


所以当"isUTF8file:false"时

判断utf8 解码长度并且编码后长度一致，那么它还是utf8,

尝试"cp437”解码，如果长度跟utf8 一致，那么它肯定不是cp438,

最后选择 用户自己的指定编码（gbk），

**Reply by gildas-lormeau** [reply]
Please propose a pull request. I guess I fixed the bug though.

**Comment by nenge123** [comment]
这是我的想法，我并不熟悉node，我用手机上的写的，若有错误自行修正
具体函数相对file:zip.js



f is option.filenameEncoding
autoDecodeText = (u8,f)=>{
    let dd = (e,s)=>new TextDecode(e).decode(new Uint8Array(s));
     let ee = e=>new TextEncoder().encode(e);

 let u = "utf8",
utf8text= dd(u,u8),
utf8len= e(utf8text);
;
if(utf8len.length!=u8.length){
//这不是utf8 编码
cp437text = decodeCP437(u8),
cp437len= e(cp437text);

if(f&&u!=f){
let usertext= dd(f,u8),
    userlen= e(utf8text);
   if(userlen.length<utf8len.length&&userlen.length<=cp437len.length)retrun usertext;
}

if(cp437len.length<utf8len.length)retrun cp437text;
//当然这里可以遍历一次常用语言

return "unknown language file";




}
retrun utf8text;


}

**Comment by nenge123** [comment]
请参考我上面的函数，这个用来解码文件名应该非常有用！

**Comment by nenge123** [comment]
我稍微修正一下
同时必须保证指定编码长度小于等于cp437len，
